File: 0101-0200/0107.Binary_Tree_Level_Order_Traversal_II.js
//0ms

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    // BFS
    if(!root) {
        return [];
    }
    let result = [];
    let queue = [root];

    while(queue.length) {
        const size = queue.length;
        let res = [];
        for(let i = 0; i < size; i++) {
            let node = queue.shift();
            res.push(node.val);
            if(node.left) queue.push(node.left);
            if(node.right) queue.push(node.right);
        }
        result.unshift(res);
    }

    return result;
};

File: 0101-0200/0107.Binary_Tree_Level_Order_Traversal_II.java
//0ms

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> l=new ArrayList<>();
        levelOrder(root,1,l);
        return l;
    }
    private void levelOrder(TreeNode root,int level,List<List<Integer>> l){
        if(root==null){
            return;
        }
        if(l.size()<level){
            l.add(0,new ArrayList<>());
        }
        l.get(l.size()-level).add(root.val);
        levelOrder(root.left,level+1,l);
        levelOrder(root.right,level+1,l);
        return;
    }
}
